### ISSUE TITLE
`ak.is_none` does not descend into option contents

### ISSUE THREAD

**Issue by agoose77** [scikit-hep/awkward#1193]
### Version of Awkward Array

1.7.0

### Description and code to reproduce

Currently, `ak.is_none` returns during the result building stage once it encounters an option type. This produces some interesting results when `ak.is_none` over nested option types:
  ```python3
  >>> index_of_index = ak.Array(
  ...     ak.layout.IndexedOptionArray64(
  ...         ak.layout.Index64(np.r_[0, 1, 2, 3]),
  ...         ak.layout.IndexedOptionArray64(
  ...             ak.layout.Index64(np.r_[0, -1, 2, 3]),
  ...             ak.layout.NumpyArray(np.r_[1, 2, 3, 4]),
  ...         ),
  ...     )
  ... )
  >>> index_of_index
  <Array [1, None, 5, 6] type='4 * ?union[?int64, ?int64]'>
  >>> ak.is_none(index_of_index)
  <Array [False, False, False, False] type='4 * bool'>
  ```

In these cases, it's visible in the type string that there is something funky going on with the layout, but the correct result would be a `NumpyArray` layout of union of bytemask arrays

It might be that the simplest solution here is to use the `transform_child_layouts` utility, and return the union of the option-layout bytemask and the result from `transform_child_layouts`, e.g.
```python3
def is_none(array, axis=0, highlevel=True, behavior=None):
    """
    Args:
        array: Data to check for missing values (None).
        axis (int): The dimension at which this operation is applied. The
            outermost dimension is `0`, followed by `1`, etc., and negative
            values count backward from the innermost: `-1` is the innermost
            dimension, `-2` is the next level up, etc.
        highlevel (bool): If True, return an #ak.Array; otherwise, return
            a low-level #ak.layout.Content subclass.
        behavior (None or dict): Custom #ak.behavior for the output array, if
            high-level.

    Returns an array whose value is True where an element of `array` is None;
    False otherwise (at a given `axis` depth).
    """

    def apply(layout, depth, posaxis):
        posaxis = layout.axis_wrap_if_negative(posaxis)
        if posaxis == depth - 1:
            nplike = ak.nplike.of(layout)
            if isinstance(layout, ak._util.optiontypes):
                mask = nplike.asarray(layout.bytemask()).view(np.bool_)

                inner_mask = ak.to_numpy(
                    ak._util.transform_child_layouts(apply, layout, depth, posaxis)
                )
                return ak.layout.NumpyArray(mask | inner_mask)

            if isinstance(
                layout,
                (
                    ak._util.unknowntypes,
                    ak._util.listtypes,
                    ak._util.recordtypes,
                    ak.layout.NumpyArray,
                ),
            ):
                return ak.layout.NumpyArray(nplike.zeros(len(layout), dtype=np.bool_))

        return ak._util.transform_child_layouts(apply, layout, depth, posaxis)

    layout = ak.operations.convert.to_layout(array)

    out = apply(layout, 1, axis)

    return ak._util.maybe_wrap_like(out, array, behavior, highlevel)
```

This feels a bit clunky, but it makes the point :)

Alternatively, we could just call `ak.packed` here, which will ensure that merge-able layouts are merged. This solution won't fix other cases, though, e.g.
```python3
ak.layout.IndexedOptionArray64(
    ak.layout.Index64(np.r_[0, 1, 2, 3]),
    ak.layout.UnionArray8_64(
        ak.layout.Index8(np.r_[0, 0, 1, 1]),
        ak.layout.Index64(np.r_[0, 1, 0, 1]),
        [
            ak.layout.IndexedOptionArray64(
                ak.layout.Index64(np.r_[0, -1, 2, 3]),
                ak.layout.NumpyArray(np.r_[1, 2, 3, 4]),
            ),
            ak.layout.NumpyArray(np.r_[4, 5, 6]),
        ],
    ),
)
```

Related to this: at present, multple option layouts grow the type string `??` corresponds to two nested options. Is this what we want? I would expect to see a single logical option (`any(l.optional for l in layouts_at_depth)`).

**Comment by jpivarski** [comment]
@agoose77, could you handle this one? From my perspective, it is not necessary to fix the v1 `is_none`, but if you can take the above into account for v2, that would be great. From your perspective, you might also need to fix v1 because you depend on it for your thesis (unless you have a work-around).

There's a project (https://github.com/scikit-hep/awkward-1.0/projects/4) where we can take ownership of high-level function porting. I was catching up on bug-fixes today and plan to take some of those porting tasks tomorrow, but I can avoid this one because you've thought more deeply about it and see something that needs to get fixed.

**Comment by jpivarski** [comment]
Thanks! (I linked the card to this issue.) If you're sure the above is a bug, you can switch the label, too.

**Comment by agoose77** [comment]
Depends upon #1279 to fix tests

**Comment by jpivarski** [comment]
@ioanaif and I were looking at the `is_none` code. It was a lot more complicated than I was expecting, and that was because PR #1249 expanded it to handle cases like `option[option[X]]` (or if one of those `options` is a type-invisible `IndexedArray`) and `option[union[option[X], ...]]`.

These cases are not supposed to exist in user ak.Arrays—issue #1910 proposes a guard against them, to try to ensure that we never encounter

* `IndexedArray` of `option`-type
* `option`-type of `IndexedArray`
* `IndexedArray` of `IndexedArray`
* `option`-type of `option`-type
* `union`-type of `union`-type
* `union`-type in which one of the `contents` is a `union`-type

This issue, #1193, also mentions

* `option`-type of `union`-type (specifically, the three-level combination, `option`-type of `union`-type in which one of the `contents` is a `union`-type, is what is being protected against)

as a case for `is_none` to handle, but this is also something that shouldn't be allowed to exist. Arrow has already forbidden `option`-type of `union`-type, and we're halfway there: we don't admit that it exists as a type: the `options` are always pushed down to be inside of the `unions`. See, for example,

```python
>>> array = ak.contents.ByteMaskedArray(
...     ak.index.Index(np.array([1, 1, 0, 1], np.int8)),
...     ak.from_iter([1, "two", 999, [1, 2, 3]], highlevel=False),
...     valid_when=True,
... )

>>> print(array.form.type)
union[?int64, ?string, option[var * int64]]
```

This layout has a `ByteMaskedArray` around a `UnionArray`, but the type claims that the `union` is on the outside and the `option` is on the inside. This is the type it would have if we did a round trip through Arrow (regardless of whether metadata are preserved with `extensionarray=True`/`False`):

```python
>>> ak.from_arrow(ak.to_arrow(array)).layout
<UnionArray len='4'>
    <tags><Index dtype='int8' len='4'>[0 1 0 2]</Index></tags>
    <index><Index dtype='int32' len='4'>[0 0 1 0]</Index></index>
    <content index='0'>
        <BitMaskedArray valid_when='true' lsb_order='true' len='2'>
            <mask><Index dtype='uint8' len='1'>[1]</Index></mask>
            <content><NumpyArray dtype='int64' len='2'>[  1 999]</NumpyArray></content>
        </BitMaskedArray>
    </content>
    <content index='1'>
        <BitMaskedArray valid_when='true' lsb_order='true' len='1'>
            <mask><Index dtype='uint8' len='1'>[1]</Index></mask>
            <content><ListOffsetArray len='1'>
                <parameter name='__array__'>'string'</parameter>
                <offsets><Index dtype='int64' len='2'>[0 3]</Index></offsets>
                <content><NumpyArray dtype='uint8' len='3'>
                    <parameter name='__array__'>'char'</parameter>
                    [116 119 111]
                </NumpyArray></content>
            </ListOffsetArray></content>
        </BitMaskedArray>
    </content>
    <content index='2'>
        <UnmaskedArray len='1'>
            <content><ListOffsetArray len='1'>
                <offsets><Index dtype='int64' len='2'>[0 3]</Index></offsets>
                <content><NumpyArray dtype='int64' len='3'>[1 2 3]</NumpyArray></content>
            </ListOffsetArray></content>
        </UnmaskedArray>
    </content>
</UnionArray>

>>> print(ak.from_arrow(ak.to_arrow(array)).layout.form.type)
union[?int64, ?string, option[var * int64]]
```

Issue #1910 is asking to forbid non-canonical layouts in the ak.Array constructor; the `array` above would be another non-canonical case.

So with all of these taken together, users won't _have_ arrays like the ones that inspired issue #1193 and PR #1249, and the implementation of `is_none` could go back to its simpler form.

Sorry that, at the time, I didn't seem to notice that the arrays that `is_none` is being protected against are non-canonical.

**Comment by agoose77** [comment]
Right  - my understanding was that we've developed an internal convention of only operating upon simplified arrays, but it might have always been the case and you just forgot! 

I was thinking of this issue when we recently discussed "canonical" layouts.
